[MCC-412658] MAuth V2 Protocol

cmcinnes-mdsol opened the pull request:
This PR implements the MAuth V2 (MWSV2) protocol as specified [here](https://docs.google.com/document/d/1X3X5V8LnNhKAqS53SfFJJnZe-TrjdMPw09tnO3ZpfdU/edit). To add query parameters to the string to sign I updated both the `MAuth::Rack::Request` and `MAuth::Faraday::Request` classes to include query strings in their attributes for signing. I added config variables that allow signing request with only v2 and authenticating requests with only v2. I added version logic to the `signed` and `authenticate!` methods. In the Rack middleware I added an informative error response to the 401 in the case when we only accept v2 but a request has v1 headers.

There may be some areas that could use more cleanup but I tried not to mess with the v1 code too much as mauth-client is battle tested. If there are specific places where cleanup would be useful and not too invasive please point them out.

@mdsol/team-16 
[MCC-412658](https://jira.mdsol.com/browse/MCC-412658)

Still working on some specs (including authentication ticket specs) but opening this as draft to get the PR going.

cmcinnes-mdsol commented on README.md:
```diff
@@ -233,3 +233,7 @@ When fetching public keys from MAuth, the following rules apply:
+
+## Protocol Versions
+
+The mauth V2 protocol was added as of v5.0.0. This protocol updates the string_to_sign to include query parameters, uses different authentication header names, and has a few other changes. See this document for more information: (DOC?). By default mauth client will sign requests with both V1 and V2, sign responses with only the version used in the request, authenticate requests with only the high ...
```
Is there (or will there be) a confluence doc detailing v2?

jcarres-mdsol replied:
I think @ChazzMaTazz has already something

cmcinnes-mdsol commented on lib/mauth/client.rb:
```diff
@@ -386,6 +491,49 @@ def signature_valid!(object)
+        rescue OpenSSL::PKey::PKeyError
+          raise InauthenticError, "Public key decryption of signature failed!\n#{$!.class}: #{$!.message}"
+        end
+        # TODO: time-invariant comparison instead of #== ?
```
Left this todo in here because I'm not really sure what it means. I don't see how `==` is time variant but maybe I'm missing something. Can anyone advise?

ejinotti-mdsol replied:
I don't understand it either.. maybe the intent was thinking of some way to compare without the timestamp.. but it's from a commit 7 years ago by someone no longer here so I don't see any value in keeping.

jcarres-mdsol commented on lib/mauth/client.rb:
```diff
@@ -298,48 +337,114 @@ def authentic?(object)
+          token_valid_v2!(object)
+          signature_valid_v2!(object)
+        else
+          if authentication_present_v2(object)
```
Although you have stated you do not want to change the code too much, this client file is way too big with don't know how many modules in modules.
I'd rather move the v2 methods into a different file, it would be easier to read, all in one spot.

cmcinnes-mdsol replied:
@jcarres-mdsol would you prefer splitting the v2 methods into one file or breaking up the client file by splitting off some of the modules into their own files, e.g. `authenticators/local_authenticator.rb`, `security_token_cacher.rb` etc.

ssteeg-mdsol commented on spec/middleware_spec.rb:
```diff
@@ -7,194 +7,243 @@
+
   it 'builds a mauth client if not given a mauth_client' do
-    mauth_config = {:mauth_baseurl => 'http://mauth', :mauth_api_version => 'v1'}
+    mauth_config = {mauth_baseurl: 'http://mauth', mauth_api_version: 'v1'}
```
👏 Thanks for the cleanup

ssteeg-mdsol commented on spec/mauth_client_spec.rb:
```diff
@@ -52,231 +66,520 @@ def (::Rails).logger
+
+    it 'correctly initializes with authenticate_with_only_v2 as true with boolean true or string "true"' do
+      [true, 'true'].each do |authenticate_with_only_v2|
+        [{:authenticate_with_only_v2 => authenticate_with_only_v2}, {'authenticate_with_only_v2' => authenticate_with_only_v2}].each do |config|
```
I realize there's a lot of existing style violations in this file, but for new changes, could you double check the style? For example:
`{:authenticate_with_only_v2 => authenticate_with_only_v2}`
`{ authenticate_with_only_v2: authenticate_with_only_v2 }`

ssteeg-mdsol commented on spec/mauth_client_spec.rb:
```diff
@@ -52,231 +66,520 @@ def (::Rails).logger
   describe '#signed' do
-    before { @request = TestSignableRequest.new(:verb => 'PUT', :request_url => '/', :body => 'himom') }
+    it 'adds only X-MWS-Time and X-MWS-Authentication headers when signing with v1 override' do
+      expect(v1_signed_req.headers.keys).to include('X-MWS-Authentication')
```
I believe these statements can be combined, per https://relishapp.com/rspec/rspec-expectations/docs/built-in-matchers/include-matcher:
`expect(v1_signed_req.headers.keys).to include('X-MWS-Authentication', 'X-MWS-Time')`
Edit: Maybe these changes just being copied over? 😅 Bit tricky to review the spec file, maybe Jordi's comment about splitting into separate files would help here too

jcarres-mdsol replied:
yeah, then at least the new files can be rubocopped

jcarres-mdsol commented on lib/mauth/dice_bag/mauth.yml.dice:
```diff
@@ -5,6 +5,8 @@ common: &common
   mauth_api_version: v1
   app_uuid: <%= configured.mauth_app_uuid! || 'fb17460e-9868-11e1-8399-0090f5ccb4d3' %>
   private_key_file: config/mauth_key
+  authenticate_with_only_v2: <%= configured.authenticate_with_only_v2 || 'false' %>
```
The effect and usecase of these configuration options probably need to be explained in the README or other doc.
At least would need to address when should I turn this on. Or if should I wait till the client updates the template?

jcarres-mdsol commented on lib/mauth/authenticators/authenticator.rb:
```diff
@@ -0,0 +1,135 @@
+      def authenticate!(object)
+        if authenticate_with_only_v2? && authentication_present_v1(object)
+          msg = 'This service requires mAuth v2 mcc-authentication header but only v1 x-mws-authentication is present'
+          raise MAuth::MissingV2Error, msg
```
This is not rescued above while the other types are, is that on puporse?

cmcinnes-mdsol replied:
It was so that I could rescue `MAuth::MissingV2Error` [here](https://github.com/mdsol/mauth-client-ruby/pull/25/files#diff-f77b06d19ecda5de902d19ad6367fb8dR28) in the rack middleware and respond appropriately. I wanted to avoid replicating the logic in the middleware of when to raise a missing v2 error but I guess `authenticate?` should really only respond true or false.

cmcinnes-mdsol replied:
Updated to rescue `MAuth::MissingV2Error`s and added logic to rack middleware for when we should return the missing v2 response.

cmcinnes-mdsol commented:
@mdsol/team-16 Providing some info about what has changed, since the diff has gotten out of control: 

I split four modules off from the `mauth_client.rb` file and split there specs out into their own files too:
V2 methods were added in `authenticators/authenticator`, `authenticators/remote_request_authenticator.rb`, `authenticators/local_authenticator`, and `singer.rb`. The V1 methods in these files have not changed.

`security_token_cacher` and its specs were also moved to a separate file but not changed.

Would have been better to put the reorganization in a separate pr: lesson learned.

cmcinnes-mdsol commented:
Ok @mdsol/team-16, I just reverted the reorg to clean up the diff. If we still want some organization I will open another PR after this one is merged.

jcarres-mdsol commented on lib/mauth/faraday.rb:
```diff
@@ -36,9 +36,16 @@ def initialize(request_env)
-        @attributes_for_signing ||= { verb: @request_env[:method].to_s.upcase, request_url: request_url, body: @request_env[:body] }
+        @attributes_for_signing ||= begin
+          request_url = @request_env[:url].path
+          request_url = '/' if request_url.empty?
```
Kind of weird it is called `request_url` when it is only the path
Also what about 
`request_path = @request_env[:url].path.empty? '/' : @request_env[:url].path` instead of two lines

cmcinnes-mdsol replied:
Agreed that `request_path` would be better. My concern though was that some people may be manually signing requests as detailed in the README [here](https://github.com/mdsol/mauth-client-ruby#other-request-and-response-signing). Then updating mauth-client would break their usage until they change from `request_url` to `request_path`.

jcarres-mdsol replied:
I think I understand now.
 There is this method called `attributes_for_signing` here but in fact is more like `potential_attributes_we_may_use_for_signing`. Each version will somehow filter the ones not needed

jcarres-mdsol commented on lib/mauth/faraday.rb:
```diff
@@ -68,6 +75,14 @@ def x_mws_time
         @response_env[:response_headers]['x-mws-authentication']
       end
+
+      def mcc_time
```
where is this used?

cmcinnes-mdsol replied:
it's used in the v2 time_validation and signature validation in `client.rb`.

jcarres-mdsol replied:
I know I'm saying the same thing again but as is now we are adding v2 only methods to this class, which force us to add v2 methods or even mixed v1-v2 methods to the Signed module. And I guess the same happens with authorization. I think we could have different classes `ResponseV2` for instance instead

cmcinnes-mdsol replied:
@mdsol/team-16 I talked about creating `ResponseV2` and `RequestV2` classes and only including the relevant methods in those class with @jcarres-mdsol. I think it would improve readability and maintainability in some files. One problem is that during the conversion process while we're signing with both V1 and V2 we will have to instantiate two request objects every time we sign. We'll also have to move the logic of deciding which protocol to sign with into both the Faraday and rack middlewares. Thoughts?

ejennings-mdsol replied:
@cmcinnes-mdsol @jcarres-mdsol In theory I like the idea of creating `ResponseV2` and `RequestV2` (and also `ResponseV1` and `RequestV1` classes and having common code elsewhere -- I also like the idea of moving the `Signer` methods out of the `Client` so we can just do something like `request.authentic?`, `response.sign!`, `request.sign!` or `response.authentic?` in the middlewares and have those objects delegate to implementations of the proper version of the protocol).  In practice, I am concerned about the risks of changing the code too much and spending additional time on this now.  While the existing code may not be as clear as we'd like it and the files too big, it has been running in production for several years and I worry that a big refactor will introduce bugs that are hard to find.  Or do you think the additions that have been made already are already very disruptive to the V1 code path?

As for creating two objects with every signature, I think we would need a `Benchmark` to know if that causes much slower performance.  For memory usage due to multiple objects,  you could try something like tje `memory_profiler` gem.

If we do go ahead with this refactor, I would prefer to see it in a separate PR, with the hope that the specs that were added here would not have to change too much to just pass after the refactor. 🤞 Having a separate PR also might facilitate doing the benchmarking and memory profiling comparison.

ejinotti-mdsol replied:
FWIW I am pretty much in agreement with @ejennings-mdsol.. been trying to wrap my head around this PR for days without much success.

There is much that I hate about this gem, with the client file in particular being an abomination: not actually a client but rather just a hideous way of gluing together signing/authenticating with config.

I am not a huge fan of the OO stuff in here either, so I'd be ok with just slapping all of the V2 functionality onto the same `Request` object for now.  In a (later / other story) reorganization I'd look to get rid of the object entirely and just have two modules for signing and authenticating that expose one function each, then let the respective middlewares worry about how/where to attach those results (since that is their business anyway).

I think the various requirements of the story, in particular the "only" flags, may be causing this PR to be so muddled.  Perhaps we could split in PRs like (1) add the basic v2 stuff and always sign with / accept both, (2) add the flags, (3) add documentation?

jcarres-mdsol commented on lib/mauth/faraday.rb:
```diff
@@ -36,9 +36,16 @@ def initialize(request_env)
+            verb: @request_env[:method].to_s.upcase,
+            request_url: request_url,
+            body: @request_env[:body],
+            query_string: @request_env[:url].query
```
So we add the query string always ? How would this interact with a receiver with and old mauth client?

cmcinnes-mdsol replied:
Not sure I understand your question. For v1 requests `query_string` will be ignored because it's not in the `SIGNATURE_COMPONENTS` constant in `MAuth::Request` class.

jcarres-mdsol replied:
I think I understand now.
 There is this method called `attributes_for_signing` here but in fact is more like `potential_attributes_we_may_use_for_signing`. Each version will somehow filter the ones not needed

jcarres-mdsol commented on lib/mauth/rack.rb:
```diff
@@ -70,11 +84,20 @@ def should_authenticate?(env)
+    # signs outgoing responses with only the protocol used to sign the request.
     class ResponseSigner < MAuth::Middleware
       def call(env)
+        # QUESTION does instaniate a new request object, rather than doing a check for the headers here, add too much overhead?
```
I think once #24  is merged, it should be expanded to compare v1 and v2 authorization and signing. 
Also by running it on develop vs this branch you could check if performance degraded

jcarres-mdsol commented on lib/mauth/client.rb:
```diff
@@ -265,19 +282,40 @@ def signed(object, attributes = {})
+          signed_headers_v1(object, attributes)
+        elsif sign_requests_with_only_v2? || v2_only_override # override used when signing responses to requests with the v2 protocol when SIGN_REQUESTS_WITH_ONLY_V2 is false
+          signed_headers_v2(object, attributes)
+        else # by default sign with both the v1 and v2 protocol
```
so when `v2_only_override: true` is set then that means we do both v1 and v2, does not sounds as much of an override

cmcinnes-mdsol replied:
when `v2_only_override: true` we only sign with v2.

jcarres-mdsol commented on lib/mauth/client.rb:
```diff
@@ -265,19 +282,40 @@ def signed(object, attributes = {})
-      # for that object.
-      def signature(object, attributes = {})
-        assert_private_key(UnableToSignError.new("mAuth client cannot sign without a private key!"))
+      def signed_headers_v2(object, attributes = {})
```
I insist v2 code should be separated. Just this diff here is quite confusing, did signature method disappear?  Etc.
This client file is a mess already, adding more to it makes it really complex to follow

cmcinnes-mdsol replied:
@jcarres-mdsol was trying to get some clarity on how to do that here: https://github.com/mdsol/mauth-client-ruby/pull/25#discussion_r292625902, but just moved v2 only methods from `client.rb` into their own file.

jcarres-mdsol commented on lib/mauth/v2.rb:
```diff
@@ -0,0 +1,134 @@
+    MWSV2_TOKEN = 'MWSV2'.freeze
+    AUTH_HEADER_DELIMITER = ';'.freeze
+
+    private
```
I don't think this is really making anything private

jcarres-mdsol commented on lib/mauth/v2.rb:
```diff
@@ -0,0 +1,134 @@
+        begin
+          actual = pubkey.public_decrypt(Base64.decode64(object.signature))
+        rescue OpenSSL::PKey::PKeyError
+          raise InauthenticError, "Public key decryption of signature failed!\n#{$!.class}: #{$!.message}"
```
use `OpenSSL::PKey::PKeyError => e` and then use `e` instead of `$!`

jcarres-mdsol commented on lib/mauth/v2.rb:
```diff
@@ -0,0 +1,134 @@
+          query_string: euresource_escape(original_query_string.to_s)
+        )
+
+        pubkey = OpenSSL::PKey::RSA.new(retrieve_public_key(object.signature_app_uuid))
```
I think calculating `actual` here could go to a different method as this one is already large

jcarres-mdsol commented on lib/mauth/v2.rb:
```diff
@@ -0,0 +1,134 @@
+          raise InauthenticError, "Public key decryption of signature failed!\n#{$!.class}: #{$!.message}"
+        end
+
+        unless expected_no_reencoding == actual || expected_euresource_style_reencoding == actual || expected_for_percent_reencoding == actual
```
longer than github line here

jcarres-mdsol commented on lib/mauth/v2.rb:
```diff
@@ -0,0 +1,134 @@
+
+      # TODO: update mAuth to be able verify authentication tickets w V2 (MCC-413109)
+      def signature_valid_v2!(object)
+        raise ArgumentError, "Remote Authenticator can only authenticate requests; received #{object.inspect}" unless object.is_a?(MAuth::Request)
```
long line

jcarres-mdsol commented on lib/mauth/v2.rb:
```diff
@@ -0,0 +1,134 @@
+      # TODO: update mAuth to be able verify authentication tickets w V2 (MCC-413109)
+      def signature_valid_v2!(object)
+        raise ArgumentError, "Remote Authenticator can only authenticate requests; received #{object.inspect}" unless object.is_a?(MAuth::Request)
+        authentication_ticket = {
```
I wonder if performance wise are we recreating these strings each time or not. I do not know if Ruby already freeze all string literals or we still need magic comments. @jfeltesse-mdsol  may know.
You could consider using constants here or even have a constant hash already with all these keys and here you dup it and fill in the values.

cmcinnes-mdsol replied:
Copied the use of string keys from the old code, but we should be able to just use symbols. I'll update.

ejennings-mdsol commented on spec/middleware_spec.rb:
```diff
@@ -7,194 +7,238 @@
+        expect(401).to eq(status)
+        expect(headers['Content-Type']).to eq('application/json')
+        expect(JSON.parse(body.join)).to eq({
+            'type' => 'errors:mauth:missing_v2',
```
Seems like too much indentation here since `({` and `})` are each on their own same line

ejennings-mdsol commented on lib/mauth/request_and_response.rb:
```diff
@@ -23,6 +39,67 @@ def string_to_sign(more_attributes)
+      end
+
+      missing_attributes.delete(:body_digest) # body may be omitted
+      if missing_attributes.any?
```
I'm wondering if there should be specs testing all permutations of missing attributes for the string to sign (for v1 and v2).  Thoughts?

cmcinnes-mdsol replied:
I think it makes sense. I'm writing specs for the `Signable` module to address this (and https://github.com/mdsol/mauth-client-ruby/pull/25#discussion_r293524108)

jcarres-mdsol replied:
I'd recommend the use of mutant https://github.com/mbj/mutant
It will force your specs to really cover all paths. We have used it for other foundational gems   where we do not want to take any chances.

mbj replied:
Feel free to ping me on my email if you wanted help with mutant use.

ejennings-mdsol commented on lib/mauth/request_and_response.rb:
```diff
@@ -23,6 +39,67 @@ def string_to_sign(more_attributes)
+
+    # percent encodes special characters, preserving character encoding
+    # identical to CGI.escape except it encodes spaces as %20%
+    # QUESTION should just use CGI.escape then gsub?
```
Seems worth writing a simple `Benchmark` to answer the question (not to be included in the PR)

ejennings-mdsol commented on lib/mauth/client.rb:
```diff
@@ -293,55 +322,92 @@ def authentic?(object)
 
-      def time_valid!(object, now = Time.now)
+      def authentication_present_v1(object)
+        !object.x_mws_authentication.nil? || object.x_mws_authentication&.match?(/\S/)
```
I would have hoped a spec would have failed here (and in the other && for || case)

cmcinnes-mdsol replied:
Yes thanks, fixed/added authentication header presence specs.

ejinotti-mdsol commented on lib/mauth/request_and_response.rb:
```diff
@@ -23,6 +39,67 @@ def string_to_sign(more_attributes)
+    # CGI.escape(string).gsub!(/\+/, '%20%')
+    def uri_escape(string)
+      encoding = string.encoding
+      string.b.gsub(/([^a-zA-Z0-9_.~-]+)/) do |m|
```
"match"? `gsub` yields each match it hits to the block i believe.

cmcinnes-mdsol replied:
Yes it stand for "match" thanks @ejinotti-mdsol. I just did a quick benchmark though and using `CGI.escape` and then `gsub` appears to be faster so I'm going to go with that.

cmcinnes-mdsol commented on spec/mauth_client_spec.rb:
```diff
@@ -52,237 +66,523 @@ def (::Rails).logger
+
+          it 'logs the mauth app uuid of the requester and requestee when they both have such uuids' do
+            v1_signed_req = client.signed(request, time: Time.now.to_i)
+            expect(authenticating_mc.logger).to receive(:info).with("Mauth-client attempting to authenticate request from app with mauth app uuid signer to app with mauth app uuid authenticator using version MWSV2.")
```
Thanks for catching, these two specs were actually signing with both v2 and v1 because they were overwriting the `let` `v1_signed_req` with their own local `v1_signed_req` that was being called without the override. Fixed and confirmed no other specs are doing this.

cmcinnes-mdsol commented on lib/mauth/client.rb:
```diff
@@ -293,55 +322,92 @@ def authentic?(object)
+        return if  (-ALLOWED_DRIFT_SECONDS..ALLOWED_DRIFT_SECONDS).cover?(now.to_i - time_signed)
+
+        msg = "Time verification failed. #{time_signed} not within #{ALLOWED_DRIFT_SECONDS} of #{now}"
+        log_mauth_not_present(object, msg)
```
I think it should `log_inauthentic` as that's what we logged for time drift violations before i added this helper.

cmcinnes-mdsol commented on lib/mauth/request_and_response.rb:
```diff
@@ -35,13 +112,27 @@ def attributes_for_signing
-        match = x_mws_authentication && x_mws_authentication.match(/\A([^ ]+) *([^:]+):([^:]+)\z/)
+        match = if mcc_authentication
+          mcc_authentication.match(
+            /\A(#{MAuth::Client::MWSV2_TOKEN}) ([^:]+):([^:]+)#{MAuth::Client::AUTH_HEADER_DELIMITER}\z/
```
If we specify the length of the app uuid, signature, or whole header then we would be throwing out invalid signatures and app uuids here before we get to the authenticator validations which could make logs and errors less helpful.
One example: if an app_uuid is too short we could end up getting 
`"mAuth service responded with 404 looking up public key for "` rather than 
`"mAuth service responded with 404 looking up public key for 123"`

cmcinnes-mdsol commented on lib/mauth/client.rb:
```diff
@@ -293,55 +322,92 @@ def authentic?(object)
-        logger.error "mAuth signature authentication failed for #{object.class}. Exception: #{e.message}"
-        raise
-      rescue UnableToAuthenticateError => e
-        logger.error "Unable to authenticate with MAuth for #{object.class}. Exception: #{e.message}"
```
It exists here: https://github.com/mdsol/mauth-client-ruby/pull/25/files#diff-42931407a1047a487049e73fad58183dR368

cmcinnes-mdsol commented on spec/mauth_client_spec.rb:
```diff
@@ -52,237 +66,523 @@ def (::Rails).logger
-          allow(signed_request).to receive(:signature_app_uuid).and_return(nil)
-          expect(@authenticating_mc.logger).to receive(:info).with("Mauth-client attempting to authenticate request from app with mauth app uuid [none provided] to app with mauth app uuid authenticator.")
-          @authenticating_mc.authentic?(signed_request) rescue nil
+        it "considers a request with no v2 or v1 headers to be inauthentic" do
```
Changed to `considers a request without v2 and v1 headers to be inauthentic` because when v1 headers are present we return an `MissingV2Error`

cmcinnes-mdsol commented on lib/mauth/request_and_response.rb:
```diff
@@ -23,6 +39,67 @@ def string_to_sign(more_attributes)
+    # sorts query string parameters by codepoint, uri encodes keys and values,
+    # and rejoins parameters into a query string
+    def encode_query_string(q_string)
+      q_string.split('&').sort.map do |part|
```
@ejennings-mdsol added a spec [here](https://github.com/mdsol/mauth-client-ruby/pull/25/files#diff-ddac022f30828525f6f8862f820e94feR628) and specs for the `Signable` module including some sorting of multi byte utf-8 chars.

ejinotti-mdsol commented on lib/mauth/rack.rb:
```diff
@@ -70,11 +86,20 @@ def should_authenticate?(env)
     class ResponseSigner < MAuth::Middleware
       def call(env)
+        # QUESTION does instaniate a new request object, rather than doing a check for the headers here, add too much overhead?
+        versioned_token = MAuth::Rack::Request.new(env).signature_token
```
We shouldn't need to rebuild the request object here.  Instead, I believe we can just stick something into the env in the `RequestAuthenticator` middleware to let us know here if we are on V1 or V2.

ejinotti-mdsol commented on lib/mauth/rack.rb:
```diff
@@ -70,11 +86,20 @@ def should_authenticate?(env)
+          if versioned_token == MAuth::Client::MWSV2_TOKEN
+            mauth_client.signed(MAuth::Rack::Response.new(*unsigned_response), v2_only_override: true)
+          else
+            mauth_client.signed(MAuth::Rack::Response.new(*unsigned_response), v1_only_override: true)
```
Not a fan of these override flags.  I'd rather see just two different function on the client like `signed_v1` and `signed_v2`.

ejinotti-mdsol replied:
With that and the above about sticking the version in env, think this could be something cleaner like:
```ruby
method = env['mauth.request_version'] == 2 ? :signed_v2 : :signed_v1
response = mauth_client.send(method, MAuth::Rack::Response.new(*unsigned_response))
response.status_headers_body
```

ejinotti-mdsol commented on lib/mauth/client.rb:
```diff
@@ -293,55 +320,95 @@ def authentic?(object)
+      end
+
+      def authentication_present_v1?(object)
+        !(object.x_mws_authentication.nil? || object.x_mws_authentication !~ /\S/)
```
```ruby
!object.x_mws_authentication.to_s.empty?
```

cmcinnes-mdsol replied:
@ejinotti-mdsol that would return true if authentication is a string of only whitespace (`'   '`) which differs from the old behavior.

ejinotti-mdsol replied:
add in a `.strip` then?

ejinotti-mdsol commented on lib/mauth/v2.rb:
```diff
@@ -0,0 +1,142 @@
+      end
+
+      def authentication_present_v2?(object)
+        !(object.mcc_authentication.nil? || object.mcc_authentication !~ /\S/)
```
```ruby
!object.mcc_authentication.to_s.empty?
```

ejinotti-mdsol commented on lib/mauth/client.rb:
```diff
@@ -293,55 +320,95 @@ def authentic?(object)
-      rescue UnableToAuthenticateError => e
-        logger.error "Unable to authenticate with MAuth for #{object.class}. Exception: #{e.message}"
-        raise
+        if authentication_present_v2?(object)
```
Correct me if I'm wrong here, but the "object" being passed to this function should be either a `MAuth::Faraday::Response` or a `MAuth::Rack::Request`?

If that is correct, then I think we could make this logic and error throwing simpler by defining a `version` function on those two classes.

Also, note for future refactor: it would prob be much nicer to have two base classes called `Incoming` and `Outgoing` as they are essentially the inverse with respect to signing vs authenticating for rack and faraday and we could save some duplicate functionality.

cmcinnes-mdsol replied:
I put a `#protocol_version` method in the `Signed` module in `request_and_response.rb`. The `Signed` module serves basically as a place to put `Incoming` methods.

ejinotti-mdsol commented on lib/mauth/client.rb:
```diff
@@ -293,55 +320,95 @@ def authentic?(object)
-      def time_valid!(object, now = Time.now)
+      def log_unable_to_authenticate(message)
+        logger.error("Unable to authenticate with MAuth. Exception #{message}")
+      end
```
this function is only used it one place?  it's also just one line so doesn't seem worthy of being a function?

ejinotti-mdsol commented on lib/mauth/client.rb:
```diff
@@ -293,55 +320,95 @@ def authentic?(object)
+
+      def log_mauth_not_present(object, message)
+        logger.warn("mAuth signature not present on #{object.class}. Exception: #{message}")
+      end
```
this function does not seem to be used anywhere?

ejinotti-mdsol commented on lib/mauth/request_and_response.rb:
```diff
@@ -4,15 +4,31 @@ module MAuth
+    #     resource_url_path (no host, port or query string; first "/" is included) + <LF> +
+    #     request_body + <LF> +
+    #     app_uuid + <LF> +
+    #     current_seconds_since_epoch + <LF> +
```
Is there really a trailing `<LF>` here?

ejinotti-mdsol commented on lib/mauth/request_and_response.rb:
```diff
@@ -57,17 +149,36 @@ def signature_token
+      elsif !x_mws_authentication.to_s.strip.empty?
+        1
+      end
+    end
```
would it be better to just examine `signature_token` for this function? i.e. if it's v1 token => 1, if v2 => 2, else nil?

cmcinnes-mdsol replied:
@ejinotti-mdsol  that's how I had it at first but we would lose some clarity in our logging (or the `authenticate!` conditional would have to get more complicated).
For example if we had a req with the header `X-MWS-Authentication` (v1) and value `'MWSV2 123;signature'` with the V2 token. Then we would try authenticating with v2 and fail because of the mismatch in signature but I think we should be authenticating with V1 and failing because of the wrong token.

ejinotti-mdsol replied:
Ah I see, header key mismatched with the token provided in the value.. hmmmmmmmm

Is it a requirement of the story that we have all of these various error cases to tell them exactly what they did wrong?  Like gave v1 key with v2 value, v2 key with v1 value, missing header entirely, etc?

Or could we instead encapsulate all of that checking here in this function like:
```ruby
if has_v1_key? && valid_v1_header_value?
  1
elsif has_v2_key? && valid_v2_header_value?
  2
else
  nil # and somewhere down the line this result in a generic you-did-something-wrong-lol error response?
end
```

cmcinnes-mdsol replied:
@ejinotti-mdsol  We have a different informative logging messages for all these scenarios but to an outside consumer any of these cases look identical (inauthentic), except for the case when an incoming request is a V1 (`X-MWS-Authentication`) header present, no v2 header present, and the `v2_authenticate_only` flag is true. (in this case the rack middleware will respond with an missing v2 error). 

The other complexity is that when the headers are missing altogether we want to raise `MAuthNotPresent` and log a warning instead of an error so that we don't spam error logs for services that deliberately don't include the header. (This was the subject of #23). 

The old authentication flow was: 
1) check for the presence of the header, log and raise if it's not there
1) check for the presence and validity of the time header, log and raise if it's not there 
1) check validity of the token, log and raise if it's not there
1) check the validity of the signature, log and raise if it's not there 

I tried to stay close to this by making the presence of the v1 or v2 header the outer conditional in `#authenticate`


We could encapsulate all of that checking here in this `protocol_versions` function as you said but I worry that would make the logging less clear for these cases of mismatched token and header name, which may be a pain point during the conversion process.

ejinotti-mdsol replied:
Hmm, ok yes I guess there is not much more to do then.  Thanks!

ejinotti-mdsol commented on lib/mauth/rack.rb:
```diff
@@ -70,12 +86,23 @@ def should_authenticate?(env)
+            # if no protocol was supplied then use `signed` which either signs
+            # with both protocol versions or only v2
+            :signed
+          end
```
is this a case that actually occurs?  like if we get a request with no mauth header whatsoever, we are still signing the error response back to them that says missing header?

cmcinnes-mdsol replied:
Was actually just testing that out and now we don't sign the "Unauthorized" responses so I'll take out this else.

cmcinnes-mdsol replied:
@ejinotti-mdsol fyi added this back in because this case can be hit:
One example i ran into while local testing the new client with eureka: eureka does not authenticate requests locally by default but still signs them.

ejinotti-mdsol commented on lib/mauth/rack.rb:
```diff
@@ -15,19 +15,23 @@ module Rack
+
+        begin
+          if mauth_client.authentic?(mauth_request)
+            @app.call(env.merge('mauth.app_uuid' => mauth_request.signature_app_uuid, 'mauth.authentic' => true))
```
not you, but maybe a slight perf increase with we do an in-place `.merge!` here on 29 or just add a couple lines writing directly to the `env` hash?

jcarres-mdsol commented:
I think we should also add v2 performance test to https://github.com/mdsol/mauth-client-ruby/blob/master/Rakefile#L31 
This PR is already big so it would be ok to do in another one.

ejennings-mdsol commented on spec/mauth_client_spec.rb:
```diff
@@ -52,237 +66,595 @@ def (::Rails).logger
+          end
+
+          it 'considers a signed request with multi-byte UTF-8 characters in the query string to be authentic' do
+            qs = 'prm=パ&prm=개'
```
What is this doing?

cmcinnes-mdsol replied:
Whoops: changed to `considers a signed request with repeated query param keys with multi-byte UTF-8 character values to be authentic` and updated request to use this query string. This was a spec to test sorting by codepoint of values if keys are the same.

cmcinnes-mdsol commented on spec/mauth_client_spec.rb:
```diff
@@ -52,237 +66,595 @@ def (::Rails).logger
+            signed_request = client.signed(request, time: Time.now.to_i + time_offset)
+            message = "expected request signed at #{time_offset} seconds to #{authentic ? "" : "not"} be authentic"
+            if authentic
+              expect(authenticating_mc.authentic?(signed_request)).to be_truthy, message
```
Copypasta from old specs but as for `message`: the specs fail with that message if this expectation fails. Helpful in a spec that runs in a loop like this so we know which value of `time_offset` failed.
